File: alljoyn_core/src/winrt/binding/src/AuthListener.cc
/******************************************************************************
 *
 * Copyright (c) 2011-2012, AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 *****************************************************************************/

#include "AuthListener.h"

#include <qcc/String.h>
#include <qcc/winrt/utility.h>
#include <Message.h>
#include <BusAttachment.h>
#include <Credentials.h>
#include <ObjectReference.h>
#include <AllJoynException.h>

namespace AllJoyn {

AuthListener::AuthListener(BusAttachment ^ bus)
{
    ::QStatus status = ER_OK;

    while (true) {
        // Check for invalid values for bus attachment
        if (nullptr == bus) {
            status = ER_BAD_ARG_1;
            break;
        }
        // Create auth listener as a managed type
        _mListener = new qcc::ManagedObj<_AuthListener>(bus);
        if (NULL == _mListener) {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Reference to T* contained in the managed interface for convenience
        _listener = &(**_mListener);
        break;
    }

    // Bubble up QStatus as a qcc exception
    if (ER_OK != status) {
        QCC_THROW_EXCEPTION(status);
    }
}

AuthListener::AuthListener(const qcc::ManagedObj<_AuthListener>* listener)
{
    ::QStatus status = ER_OK;

    while (true) {
        // Check for invalid values specified for listener
        if (NULL == listener) {
            status = ER_BAD_ARG_1;
            break;
        }
        // Attach the listener passed in
        _mListener = new qcc::ManagedObj<_AuthListener>(*listener);
        if (NULL == _mListener) {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Reference to T* contained in the managed interface for convenience
        _listener = &(**_mListener);
        break;
    }

    // Bubble up QStatus as a qcc exception
    if (ER_OK != status) {
        QCC_THROW_EXCEPTION(status);
    }
}

AuthListener::~AuthListener()
{
    // Delete the interface managed interface to reflect current use count
    if (NULL != _mListener) {
        delete _mListener;
        _mListener = NULL;
        _listener = NULL;
    }
}

void AuthListener::RequestCredentialsResponse(AuthContext ^ authContext, bool accept, Credentials ^ credentials)
{
    ::QStatus status = ER_OK;

    while (true) {
        // Check for invalid values specified in authContext
        if (nullptr == authContext) {
            status = ER_BAD_ARG_1;
            break;
        }
        // Check for invalid values specified in credentials
        if (nullptr == credentials) {
            status = ER_BAD_ARG_3;
            break;
        }
        // Get the ajn Credentials type for the actual call
        ajn::AuthListener::Credentials* creds = credentials->_credentials;
        // Call the real API with parameters
        status = _listener->RequestCredentialsResponse(authContext->_authContext, accept, *creds);
        break;
    }

    // Bubble up QStatus as a qcc exception
    if (ER_OK != status) {
        QCC_THROW_EXCEPTION(status);
    }
}

void AuthListener::VerifyCredentialsResponse(AuthContext ^ authContext, bool accept)
{
    ::QStatus status = ER_OK;

    while (true) {
        // Check for invalid values specified in authContext
        if (nullptr == authContext) {
            status = ER_BAD_ARG_1;
            break;
        }
        // Call the real API with parameters
        status = _listener->VerifyCredentialsResponse(authContext->_authContext, accept);
        break;
    }

    // Bubble up QStatus as a qcc exception
    if (ER_OK != status) {
        QCC_THROW_EXCEPTION(status);
    }
}

Windows::Foundation::EventRegistrationToken AuthListener::RequestCredentials::add(AuthListenerRequestCredentialsAsyncHandler ^ handler)
{
    // Adds a handler to the internal ref class for RequestCredentials
    return _listener->_eventsAndProperties->RequestCredentials::add(handler);
}

void AuthListener::RequestCredentials::remove(Windows::Foundation::EventRegistrationToken token)
{
    // Removes a handler to the internal ref class for RequestCredentials
    _listener->_eventsAndProperties->RequestCredentials::remove(token);
}

QStatus AuthListener::RequestCredentials::raise(Platform::String ^ authMechanism, Platform::String ^ peerName, uint16_t authCount, Platform::String ^ userName, uint16_t credMask, AuthContext ^ authContext)
{
    // Invokes a handler to the internal ref class for RequestCredentials
    return _listener->_eventsAndProperties->RequestCredentials::raise(authMechanism, peerName, authCount, userName, credMask, authContext);
}

Windows::Foundation::EventRegistrationToken AuthListener::VerifyCredentials::add(AuthListenerVerifyCredentialsAsyncHandler ^ handler)
{
    // Adds a handler to the internal ref class for VerifyCredentials
    return _listener->_eventsAndProperties->VerifyCredentials::add(handler);
}

void AuthListener::VerifyCredentials::remove(Windows::Foundation::EventRegistrationToken token)
{
    // Removes a handler to the internal ref class for VerifyCredentials
    _listener->_eventsAndProperties->VerifyCredentials::remove(token);
}

QStatus AuthListener::VerifyCredentials::raise(Platform::String ^ authMechanism, Platform::String ^ peerName, AllJoyn::Credentials ^ credentials, AuthContext ^ authContext)
{
    // Invokes a handler to the internal ref class for VerifyCredentials
    return _listener->_eventsAndProperties->VerifyCredentials::raise(authMechanism, peerName, credentials, authContext);
}

Windows::Foundation::EventRegistrationToken AuthListener::SecurityViolation::add(AuthListenerSecurityViolationHandler ^ handler)
{
    // Adds a handler to the internal ref class for SecurityViolation
    return _listener->_eventsAndProperties->SecurityViolation::add(handler);
}

void AuthListener::SecurityViolation::remove(Windows::Foundation::EventRegistrationToken token)
{
    // Removes a handler to the internal ref class for SecurityViolation
    _listener->_eventsAndProperties->SecurityViolation::remove(token);
}

void AuthListener::SecurityViolation::raise(QStatus status, Message ^ msg)
{
    // Invokes a handler to the internal ref class for SecurityViolation
    _listener->_eventsAndProperties->SecurityViolation::raise(status, msg);
}

Windows::Foundation::EventRegistrationToken AuthListener::AuthenticationComplete::add(AuthListenerAuthenticationCompleteHandler ^ handler)
{
    // Adds a handler to the internal ref class for AuthenticationComplete
    return _listener->_eventsAndProperties->AuthenticationComplete::add(handler);
}

void AuthListener::AuthenticationComplete::remove(Windows::Foundation::EventRegistrationToken token)
{
    // Removes a handler to the internal ref class for AuthenticationComplete
    _listener->_eventsAndProperties->AuthenticationComplete::remove(token);
}

void AuthListener::AuthenticationComplete::raise(Platform::String ^ authMechanism, Platform::String ^ peerName, bool success)
{
    // Invokes a handler to the internal ref class for AuthenticationComplete
    _listener->_eventsAndProperties->AuthenticationComplete::raise(authMechanism, peerName, success);
}

BusAttachment ^ AuthListener::Bus::get()
{
    // Get property accessor to the internal ref class for Bus
    return _listener->_eventsAndProperties->Bus;
}

_AuthListener::_AuthListener(BusAttachment ^ bus)
{
    ::QStatus status = ER_OK;

    while (true) {
        // Create an instance of the internal ref class
        _eventsAndProperties = ref new __AuthListener();
        // Check for failed allocation
        if (nullptr == _eventsAndProperties) {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Set a default handler to the internal ref class for RequestCredentials
        _eventsAndProperties->RequestCredentials += ref new AuthListenerRequestCredentialsAsyncHandler([&] (Platform::String ^ authMechanism, Platform::String ^ peerName, uint16_t authCount, Platform::String ^ userName, uint16_t credMask, AuthContext ^ authContext)->QStatus {
                                                                                                           return DefaultAuthListenerRequestCredentialsAsyncHandler(authMechanism, peerName, authCount, userName, credMask, authContext);
                                                                                                       });
        // Set a default handler to the internal ref class for VerifyCredentials
        _eventsAndProperties->VerifyCredentials += ref new AuthListenerVerifyCredentialsAsyncHandler([&] (Platform::String ^ authMechanism, Platform::String ^ peerName, AllJoyn::Credentials ^ credentials, AuthContext ^ authContext)->QStatus {
                                                                                                         return DefaultAuthListenerVerifyCredentialsAsyncHandler(authMechanism, peerName, credentials, authContext);
                                                                                                     });
        // Set a default handler to the internal ref class for SecurityViolation
        _eventsAndProperties->SecurityViolation += ref new AuthListenerSecurityViolationHandler([&] (AllJoyn::QStatus status, Message ^ msg) {
                                                                                                    DefaultAuthListenerSecurityViolationHandler(status, msg);
                                                                                                });
        // Set a default handler to the internal ref class for AuthenticationComplete
        _eventsAndProperties->AuthenticationComplete += ref new AuthListenerAuthenticationCompleteHandler([&] (Platform::String ^ authMechanism, Platform::String ^ peerName, bool success) {
                                                                                                              DefaultAuthListenerAuthenticationCompleteHandler(authMechanism, peerName, success);
                                                                                                          });
        // Set the internal ref class bus attachment value
        _eventsAndProperties->Bus = bus;
        break;
    }

    // Bubble up QStatus as a qcc exception
    if (ER_OK != status) {
        QCC_THROW_EXCEPTION(status);
    }
}

_AuthListener::~_AuthListener()
{
    _eventsAndProperties = nullptr;
}

QStatus _AuthListener::DefaultAuthListenerRequestCredentialsAsyncHandler(Platform::String ^ authMechanism, Platform::String ^ peerName, uint16_t authCount, Platform::String ^ userName, uint16_t credMask, AuthContext ^ authContext)
{
    ::QStatus status = ER_OK;

    while (true) {
        // Convert the authMechnaism value to a qcc::String
        qcc::String strAuthMechanism = PlatformToMultibyteString(authMechanism);
        // Check for failed conversion
        if (nullptr != authMechanism && strAuthMechanism.empty()) {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Convert the peerName value to a qcc::String
        qcc::String strPeerName = PlatformToMultibyteString(peerName);
        // Check for failed conversion
        if (nullptr != peerName && strPeerName.empty()) {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Convert the userName value to a qcc::String
        qcc::String strUserName = PlatformToMultibyteString(userName);
        // Check for failed conversion
        if (nullptr != userName && strUserName.empty()) {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Call the real API with parameters
        status = ajn::AuthListener::RequestCredentialsAsync(strAuthMechanism.c_str(), strPeerName.c_str(), authCount, strUserName.c_str(), credMask, authContext->_authContext);
        break;
    }

    return (AllJoyn::QStatus)status;
}

QStatus _AuthListener::DefaultAuthListenerVerifyCredentialsAsyncHandler(Platform::String ^ authMechanism, Platform::String ^ peerName, AllJoyn::Credentials ^ credentials, AuthContext ^ authContext)
{
    ::QStatus status = ER_OK;

    while (true) {
        // Convert the authMechnaism value to a qcc::String
        qcc::String strAuthMechanism = PlatformToMultibyteString(authMechanism);
        // Check for failed conversion
        if (nullptr != authMechanism && strAuthMechanism.empty()) {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Convert the peerName value to a qcc::String
        qcc::String strPeerName = PlatformToMultibyteString(peerName);
        // Check for failed conversion
        if (nullptr != peerName && strPeerName.empty()) {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Get the ajn Credentials type for the actual call
        ajn::AuthListener::Credentials* creds = credentials->_credentials;
        // Call the real API with parameters
        status = ajn::AuthListener::VerifyCredentialsAsync(strAuthMechanism.c_str(), strPeerName.c_str(), *creds, authContext->_authContext);
        break;
    }

    return (AllJoyn::QStatus)status;
}

void _AuthListener::DefaultAuthListenerSecurityViolationHandler(AllJoyn::QStatus status, Message ^ msg)
{
    //Get the ajn Message type for the actual call
    ajn::Message* m = *(msg->_message);
    // Call the real API with parameters
    ajn::AuthListener::SecurityViolation((::QStatus)(int)status, *m);
}

void _AuthListener::DefaultAuthListenerAuthenticationCompleteHandler(Platform::String ^ authMechanism, Platform::String ^ peerName, bool success)
{
    // For virtual pure functions, do nothing
}

::QStatus _AuthListener::RequestCredentialsAsync(const char* authMechanism, const char* peerName, uint16_t authCount, const char* userName, uint16_t credMask, void* authContext)
{
    ::QStatus status = ER_OK;

    while (true) {
        // Convert the authMechnaism value to a Platform::String
        Platform::String ^ strAuthMechanism = MultibyteToPlatformString(authMechanism);
        // Check for failed conversion
        if (nullptr == strAuthMechanism && authMechanism != NULL && authMechanism[0] != '\0') {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Convert the peerName value to a Platform::String
        Platform::String ^ strPeerName = MultibyteToPlatformString(peerName);
        // Check for failed conversion
        if (nullptr == strPeerName && peerName != NULL && peerName[0] != '\0') {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Convert the userName value to a Platform::String
        Platform::String ^ strUserName = MultibyteToPlatformString(userName);
        // Check for failed conversion
        if (nullptr == strUserName && userName != NULL && userName[0] != '\0') {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Create auth context from unmanaged type
        AuthContext ^ context = ref new AuthContext(authContext);
        // Check for failed allocation
        if (nullptr == context) {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Invoke the callback for RequestCredentials through the dispatcher
        _eventsAndProperties->Bus->_busAttachment->DispatchCallback(ref new Windows::UI::Core::DispatchedHandler([&]() {
                                                                                                                     status = (::QStatus)_eventsAndProperties->RequestCredentials(strAuthMechanism, strPeerName, authCount, strUserName, credMask, context);
                                                                                                                 }));
        break;
    }

    // Bubble up QStatus as a qcc exception
    if (ER_OK != status) {
        QCC_THROW_EXCEPTION(status);
    }

    return status;
}

::QStatus _AuthListener::VerifyCredentialsAsync(const char* authMechanism, const char* peerName, const Credentials& credentials, void* authContext)
{
    ::QStatus status = ER_OK;

    while (true) {
        // Convert the authMechnaism value to a Platform::String
        Platform::String ^ strAuthMechanism = MultibyteToPlatformString(authMechanism);
        // Check for failed conversion
        if (nullptr == strAuthMechanism && authMechanism != NULL && authMechanism[0] != '\0') {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Convert the peerName value to a Platform::String
        Platform::String ^ strPeerName = MultibyteToPlatformString(peerName);
        // Check for failed conversion
        if (nullptr == strPeerName && peerName != NULL && peerName[0] != '\0') {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Create credentials from unmanaged type
        AllJoyn::Credentials ^ cred = ref new AllJoyn::Credentials(&credentials);
        // Check for failed allocation
        if (nullptr == cred) {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Create auth context from unmanaged type
        AuthContext ^ context = ref new AuthContext(authContext);
        // Check for failed allocation
        if (nullptr == context) {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Invoke the callback for VerifyCredentials through the dispatcher
        _eventsAndProperties->Bus->_busAttachment->DispatchCallback(ref new Windows::UI::Core::DispatchedHandler([&]() {
                                                                                                                     status = (::QStatus)_eventsAndProperties->VerifyCredentials(strAuthMechanism, strPeerName, cred, context);
                                                                                                                 }));
        break;
    }

    // Bubble up QStatus as a qcc exception
    if (ER_OK != status) {
        QCC_THROW_EXCEPTION(status);
    }

    return status;
}

void _AuthListener::SecurityViolation(::QStatus status, const ajn::Message& msg)
{
    while (true) {
        // Create message from unmanaged type
        Message ^ message = ref new Message(&msg);
        // Check for failed allocation
        if (nullptr == message) {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Invoke the callback for SecurityViolation through the dispatcher
        _eventsAndProperties->Bus->_busAttachment->DispatchCallback(ref new Windows::UI::Core::DispatchedHandler([&]() {
                                                                                                                     _eventsAndProperties->SecurityViolation((QStatus)(int)status, message);
                                                                                                                 }));
        break;
    }

    // Bubble up QStatus as a qcc exception
    if (ER_OK != status) {
        QCC_THROW_EXCEPTION(status);
    }
}

void _AuthListener::AuthenticationComplete(const char* authMechanism, const char* peerName, bool success)
{
    ::QStatus status = ER_OK;

    while (true) {
        // Convert the authMechanism value to a Platform::String
        Platform::String ^ strAuthMechanism = MultibyteToPlatformString(authMechanism);
        // Check for failed conversion
        if (nullptr == strAuthMechanism && authMechanism != NULL && authMechanism[0] != '\0') {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Convert the peerName value to a Platform::String
        Platform::String ^ strPeerName = MultibyteToPlatformString(peerName);
        // Check for failed conversion
        if (nullptr == strPeerName && peerName != NULL && peerName[0] != '\0') {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Invoke the callback for AuthenticationComplete through the dispatcher
        _eventsAndProperties->Bus->_busAttachment->DispatchCallback(ref new Windows::UI::Core::DispatchedHandler([&]() {
                                                                                                                     _eventsAndProperties->AuthenticationComplete(strAuthMechanism, strPeerName, success);
                                                                                                                 }));
        break;
    }

    // Bubble up QStatus as a qcc exception
    if (ER_OK != status) {
        QCC_THROW_EXCEPTION(status);
    }
}

__AuthListener::__AuthListener()
{
    Bus = nullptr;
}

__AuthListener::~__AuthListener()
{
    Bus = nullptr;
}

}

File: alljoyn_java/src/org/alljoyn/bus/AuthListener.java
/*
 * Copyright (c) 2009-2011, AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

package org.alljoyn.bus;

/**
 * Authentication listeners are responsible for handling AllJoyn authentication
 * requests.
 *
 * Listener objects are the Java objects that handle notification events and are
 * called from AllJoyn in the context of one of its threads.  All listener
 * objects are expected to be multithread safe (MT-Safe) between construction
 * and destruction.
 *
 * This rule extends to other objects accessed during processing of
 * notifications.  For example, it is a programming error to allow a notifiation
 * method to update a collection in another object without serializing access
 * to the collection.
 *
 * The important consideration in this case is that as soon as one sets up a
 * listener object to receive notifications from AllJoyn, one is implicitly
 * dealing with multithreaded code.
 *
 * Since listener objects generally run in the context of the AllJoyn thread
 * which manages reception of events, If a blocking AllJoyn call is made in
 * the context of a notification, the necessary and sufficient conditions for
 * deadlock are established.
 *
 * The important consideration in this case is that when one receives a
 * notification from AllJoyn, that notification is executing in the context of
 * an AllJoyn thread.  If one makes a blocking call back into AllJoyn on that
 * thread, a deadlock cycle is likely, and if this happens your bus attachment
 * receive thread will deadlock (with itself).  The deadlock is typically broken
 * after a bus timeout eventually happens.
 */
public interface AuthListener {

    /** Authentication credentials set via authentication requests. */
    class Credentials {
        byte[] password;
        String userName;
        String certificateChain;
        String privateKey;
        byte[] logonEntry;
        Integer expiration;

        Credentials() {}
    }

    /** Authentication request. */
    class AuthRequest {
        protected Credentials credentials;

        private AuthRequest() {}
    }

    /** Authentication request for a password, pincode, or passphrase. */
    class PasswordRequest extends AuthRequest {
        private boolean isNew;
        private boolean isOneTime;

        PasswordRequest(Credentials credentials, boolean isNew, boolean isOneTime) {
            this.credentials = credentials;
            this.isNew = isNew;
            this.isOneTime = isOneTime;
        }

        /**
         * Indicates request is for a newly created password.
         *
         * @return {@code true} if request is for a newly created password
         */
        public boolean isNewPassword() {
            return isNew;
        }

        /**
         * Indicates a request is for a one time use password.
         *
         * @return {@code true} if request is for a one time use password
         */
        public boolean isOneTimePassword() {
            return isOneTime;
        }

        /**
         * Sets a requested password, pincode, or passphrase.
         *
         * @param password the password to set
         */
        public void setPassword(char[] password) {
            credentials.password = BusAttachment.encode(password);
        }
    }

    /** Authentication request for a user name. */
    class UserNameRequest extends AuthRequest {

        UserNameRequest(Credentials credentials) {
            this.credentials = credentials;
        }

        /**
         * Sets a requested user name.
         *
         * @param userName the user name to set
         */
        public void setUserName(String userName) {
            credentials.userName = userName;
        }
    }

    /** Authentication request for a chain of PEM-encoded X509 certificates. */
    class CertificateRequest extends AuthRequest {

        CertificateRequest(Credentials credentials) {
            this.credentials = credentials;
        }

        /**
         * Sets a requested public key certificate chain. The certificate must
         * be PEM encoded.
         *
         * @param certificateChain the certificate to chain to set
         */
        public void setCertificateChain(String certificateChain) {
            credentials.certificateChain = certificateChain;
        }
    }

    /** Authentication request for a PEM encoded private key. */
    class PrivateKeyRequest extends AuthRequest {

        PrivateKeyRequest(Credentials credentials) {
            this.credentials = credentials;
        }

        /**
         * Sets a requested private key. The private key must be PEM encoded and
         * may be encrypted.
         *
         * @param privateKey the private key to set
         */
        public void setPrivateKey(String privateKey) {
            credentials.privateKey = privateKey;
        }
    }

    /** Authentication request for a logon entry. */
    class LogonEntryRequest extends AuthRequest {

        LogonEntryRequest(Credentials credentials) {
            this.credentials = credentials;
        }

        /**
         * Sets a logon entry. For example for the Secure Remote
         * Password protocol in RFC 5054, a logon entry encodes the
         * N,g, s and v parameters. An SRP logon entry has the form
         * N:g:s:v where N,g,s, and v are ASCII encoded hexadecimal
         * strings and are seperated by colons.
         *
         * @param logonEntry the logon entry to set
         */
        public void setLogonEntry(char[] logonEntry) {
            credentials.logonEntry = BusAttachment.encode(logonEntry);
        }
    }

    /** Authentication request to set an expiration time for the credentials. */
    class ExpirationRequest extends AuthRequest {

        ExpirationRequest(Credentials credentials) {
            this.credentials = credentials;
        }

        /**
         * Sets an expiration time in seconds relative to the current time for
         * the credentials. This value is optional and can be set on any
         * response to a credentials request. After the specified expiration
         * time has elapsed any secret keys based on the provided credentials
         * are invalidated and a new authentication exchange will be
         * required. If an expiration is not set the default expiration time for
         * the requested authentication mechanism is used.
         *
         * @param expiration the expiration time in seconds
         */
        public void setExpiration(int expiration) {
            credentials.expiration = expiration;
        }
    }

    /**
     * Authentication request for verification of a certificate chain from a
     * remote peer.
     */
    class VerifyRequest extends AuthRequest {
        private String certificateChain;

        VerifyRequest(String certificateChain) {
            this.certificateChain = certificateChain;
        }

        /**
         * Gets the PEM encoded X509 certificate chain to verify.
         *
         * @return an X509 certificate chain
         */
        public String getCertificateChain() {
            return certificateChain;
        }
    }

    /**
     * Called by an authentication mechanism making authentication requests.
     * A count allows the listener to decide whether to allow or reject mutiple
     * authentication attempts to the same peer.
     *
     * Any implementation of this function must be multithread safe.  See the
     * class documentation for details.
     *
     * @param mechanism the name of the authentication mechanism issuing the
     *                  request
     * @param peerName  the name of the remote peer being authenticated.  On the
     *                  initiating side this will be a well-known-name for the
     *                  remote peer. On the accepting side this will be the
     *                  unique bus name for the remote peer.
     * @param count the count (starting at 1) of the number of authentication
     *              request attempts made
     * @param userName the user name for the credentials being requested.  If
     *               this is not the empty string the request is specific to the
     *               named user.
     * @param requests the requests.  The application may handle none, some, or
     *                 all of the requests.
     *
     * @return {@code true} if the request is accepted or {@code false} if the request is
     *         rejected.  If the request is rejected the authentication is
     *         complete.
     */
    boolean requested(String mechanism, String peerName, int count, String userName,
                      AuthRequest[] requests);

    /**
     * Called by the authentication engine when all authentication attempts are
     * completed.
     *
     * Any implementation of this function must be multithread safe.  See the
     * class documentation for details.
     *
     * @param mechanism the name of the authentication mechanism that was used
     *                  or an empty string if the authentication failed
     * @param peerName  the name of the remote peer being authenticated.  On the
     *                  initiating side this will be a well-known-name for the
     *                  remote peer. On the accepting side this will be the
     *                  unique bus name for the remote peer.
     * @param authenticated {@code true} if the authentication succeeded
     */
    void completed(String mechanism, String peerName, boolean authenticated);
}
